driver-core/src/main/java/com/datastax/driver/core/CodecRegistry.java

```diff
@@ -219,7 +201,65 @@ public int hashCode() {
         }
     }
 
-    private final boolean immutable;
+    /**
+     * A complexity-based weigher.
+     * Weights are computed according to the CQL type:
+     * <ol>
+     * <li>Primtive types weigh 0;
+     * <li>Collections weigh the total weight of their inner types + the weight of their level of deepness;
+     * <li>UDTs and tuples weigh the total weight of their inner types + the weight of their level of deepness, but cannot weigh less than 1;
+     * <li>Custom (non-CQL) types weigh 1.
+     * </ol>
+     * A consequence of this algorithm is that primitive types and all "shallow" collections thereof
+     * are never evicted.
+     */
+    private static class TypeCodecWeigher implements Weigher<CacheKey, TypeCodec<?>> {
```

Comment:
Should we consider making it impossible/hard to evict manually registered codecs by decorating them in a way that they can have a weight that causes them to have the same priority as default ones?  If they map to a non-collection type they'll have a 0 weight, but if it's something like an array codec that maps to a list for example, it's treated the same as other collection codecs.